### PR TITLE
adding support for sideband streaming

### PR DESCRIPTION
**Pull request by asumit**
### What does this Pull Request accomplish?

This PR adds infrastructure to add streaming support for respective driver apis.

We are adding moniker service source files directly in this repo as a stop gap to enable submission of driver specific
streaming support. We would reference the source files directly from the sideband repo as a subsequent PR and remove thse from the source in the grpc-device repo.
The streaming functionality is aided by the grpc-sideband implementation that is added as a submodule to the grpc-device repo.
This PR build is subject to succesful completion of 2 PRs on the grpc-sideband repo 
1. https://github.com/ni/grpc-sideband/pull/7 : This PR disables warnings ( being treated as an error during build in the sideband repo due to coersion of data types
   int64_t -> int and size_t -> int
2. https://github.com/ni/grpc-sideband/pull/6/files : This PR adds the option of statically building the ni_grpc_sideband lib.

grpc-device is expected to link statically against the sideband lib

The following changes are present in this PR.
1. Option INCLUDE_SIDEBAND_RDMA has been set to OFF to disable RDMA functionality in the sideband lib.
2. Option SIDEBAND_STATIC has been set to ON to enable building the sideband lib as a static lib.
3. moniker service source files (data_moniker.proto, data_moniker_service.cpp and data_moniker_service.h) have been added directly in this repo
4. The new moniker service runs on a separate thread on port 50055
4. core_server.cpp introduces the sideband read write loop that is set up to perform the read write loop on the device.
6. core_server.cpp also has RT specific code changes to add schedule affinity on different cores to get low latency on the read/ write steps.
7. Also introduced additional configuration property in the server_config.json file to set the address of the sideband service. server side needs to set the host machines IP address in there.
8. Moniker service is registered inside the core_services_registrar.cpp. 
9. Moniker service is enabled by default and not hidden under faeture toggle.
10. sideband feature is added under a feature toggle. Please add either "feature_toggles": { "sidebandserver": true } or "code_readiness": "nextrelease"  in the server_config.json file to enable this feature

### What testing has been done?

Driver specific streaming apis will be added later. hence testing is only done to ensure that the moniker service is up and running and clients are able to call BeginSidebandStream without error.

Regular driver api for daq works fine as it is.

### CODE REVIEW

File: source/server/server_configuration_parser.cpp
Comment (astarche) [reply source/server/server_configuration_parser.cpp:105]: CLARIFICATION: "disabled" is also an acceptable default for the sideband server. "disabled" is a great default. "local connections only" is an OK default. "enabled and open to remote connections" is not OK.

File: source/protobuf/data_moniker.proto
Comment (astarche) [inline source/protobuf/data_moniker.proto:1]: The description says:

> We would reference the source files directly from the sideband repo as a subsequent PR and remove thse from the source in the grpc-device repo.

What files specifically will be moved to the sideband repo? This `proto`? Will you also revert the changes to `core_server` and `server_confguration` and move that to the `grpc-sideband` repo?

Comment (astarche) [reply source/protobuf/data_moniker.proto:1]: There is more "sideband" code than I expect added in this change. If sideband is included as a submodule. Doesn't that have the proto file and server implementation?

**Review by astarche — COMMENTED** [review]
I would like to look at this again after you have answered my questions and gotten feedback from at least 1 other reviewer.

File: source/server/server_configuration_parser.cpp
Comment (asumit) [reply source/server/server_configuration_parser.cpp:105]: I will need to add "localhost" or [::1] in the default config. With this PR there is no default value for the config which means no connections will be possible. 
Only if the user provides the external ip address, a client would be able to connect to it from another system. 

I was planning to add the default config as a subsequent PR.  But I can add in this as well

File: source/server/server_configuration_parser.cpp
Comment (asumit) [reply source/server/server_configuration_parser.cpp:20]: I agree with the sideband_address name. 

regarding the port configuration, the sideband repo currently hardcodes it to [50055](https://github.com/ni/grpc-sideband/blob/b1cd688921128044054592315327b7768e7f894c/src/sideband_data.cc#L340C9-L340C50).

Would it be good idea to have another setting in the config file that has to be set to true to enable sideband (moniker service )? If enable_sideband is true in the config only then we sideband_address will make sense

File: source/protobuf/data_moniker.proto
Comment (asumit) [reply source/protobuf/data_moniker.proto:1]: Only the proto and the data_moniker service will be referenced directly by the grpc-device code. 

The core-server and server registartion has to live here since we are creating the service inside the grpc-device server.

The sideband repo code serves to 

1. provide utility class for managing the sideband communication.
2. provide moniker service implementation and its interface for clients (.proto file).

File: source/server/server_configuration_parser.cpp
Comment (astarche) [reply source/server/server_configuration_parser.cpp]: Discussed offline: use `feature_toggles` to ensure that the code is disabled by default. We can refine the disabling switch in the future if needed. When/if the disabling mechanism changes, we should ensure that the default listener is local only.

File: source/server/server_configuration_parser.cpp
Comment (astarche) [reply source/server/server_configuration_parser.cpp]: `port` can be added later, especially since this will be behind a toggle. I think it is an important feature, but is not needed "right now" at the beginning of dev.

**Review by astarche — APPROVED** [review]

File: source/server/core_server.cpp
Comment (asumit) [inline source/server/core_server.cpp:262]: @ccifra I think this code is specific to the RT use case but we would be executing this even for linux desktop. I haven't seen a side effect on linux desktop but is it something I need to think about regarding the behavior on RT vs Linux Desktop?
@astarche

File: source/server/server_configuration_parser.cpp
Comment (asumit) [reply source/server/server_configuration_parser.cpp]: Updated the code to have the sideband streaming code under a feature toggle. If we do not turn on the Feature Toggle, sideband service is not started and other ssideband related code is also not executed.

File: source/server/core_server.cpp
Comment (astarche) [reply source/server/core_server.cpp:262]: Can you put this behind a toggle check?

**Review by astarche — APPROVED** [review]

File: source/server/core_server.cpp
Comment (asumit) [reply source/server/core_server.cpp:264]: My bad, I had it locally but forgot to push the change.

File: CMakeLists.txt
Comment (doshirohan) [inline CMakeLists.txt:554]: This would change to `_GRPC_SIDEBAND` once you address other comment towards top of this file related to `USE_SUBMODULE_LIBS`

File: CMakeLists.txt
Comment (doshirohan) [inline CMakeLists.txt:607]: Move this after `${calibrationoperations_restricted_grpc_srcs}` below as per my other comment.

File: source/server/data_moniker_service.cpp
Comment (doshirohan) [inline source/server/data_moniker_service.cpp:144]: Should this return error to client instead? Is that even possible?

File: source/server/data_moniker_service.cpp
Comment (doshirohan) [reply source/server/data_moniker_service.cpp:119]: I see multiple places in this file where `std::cout` is used. I think most, if not all, of it is debugging code and should be removed.

**Review by doshirohan — APPROVED** [review]

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt]: Updated the description

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt]: Updated

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt]: Moved this under USE_SUBMODULE_LIBS and updated the below case to use find_program

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt]: Moved data_moniker.proto and removed this code.

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt]: Updated

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt]: Updated

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt]: Updated

File: source/server/core_services_registrar.cpp
Comment (amehra-ni) [reply source/server/core_services_registrar.cpp]: I have added the helper in the data_monikder_service.cpp similar to what you shared. Let me know if it's fine or not

File: source/server/data_moniker_service.cpp
Comment (amehra-ni) [reply source/server/data_moniker_service.cpp]: removed

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt]: updated

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt]: Updated

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt]: Added it there as well. We might use these later when writing tests, I guess.

File: source/server/core_server.cpp
Comment (amehra-ni) [reply source/server/core_server.cpp]: Moved it under the same mutex when starting the main server.

File: source/server/data_moniker_service.cpp
Comment (amehra-ni) [reply source/server/data_moniker_service.cpp]: Will be working on if it's feasible in a follow-up with other changes.

File: source/config/localhost_config.json
Comment (amehra-ni) [reply source/config/localhost_config.json:4]: Will look at it once Sumit is back and raise follow-up.

File: source/protobuf/data_moniker.proto
Comment (amehra-ni) [reply source/protobuf/data_moniker.proto]: Will look at it once Sumit is back and raise follow-up.

File: CMakeLists.txt
Comment (maxxboehme) [inline CMakeLists.txt:125]: If grpc-sideband was updated to using "Modern" CMake with targets. It could specify its header files locations so that components that depend on the library wouldn't need to add these themselves. See [target_include_directories](https://cmake.org/cmake/help/latest/command/target_include_directories.html)

Something like the following in [grpc-sideband/CMakeLists.txt](https://github.com/ni/grpc-sideband/blob/b1cd688921128044054592315327b7768e7f894c/CMakeLists.txt#L35)
```cmake
target_include_directories(ni_grpc_sideband
    PUBLIC
	src
        moniker_src)
```

**Review by maxxboehme — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt:125]: I've a PR out in the sideband repo for this https://github.com/ni/grpc-sideband/pull/9 which is waiting for approval.
Can we submit this change as is and we would clean it up as a follow-up when the other PR goes in?
This would help us get unblocked on other work items.

**Review by astarche — APPROVED** [review]

File: CMakeLists.txt
Comment (amehra-ni) [reply CMakeLists.txt:124]: We will be taking care of it in another user story, where we will remove the copies of data_moniker and use them directly.

File: source/server/core_server.cpp
Comment (amehra-ni) [reply source/server/core_server.cpp]: I'll update this in follow-up. As currently it's behind feature toggle, we should not worry that if someone runs into it.

**Pull request merged**
